eden/scm/edenscm/git.py

```diff
@@ -142,7 +143,25 @@ def clone(ui, url, destpath=None, update=True, pullnames=None):
 
                 if not nodes:
                     pullnames.append(update)
-
+            else:
+                b = callgitnorepo(ui, ["ls-remote", "--symref", url, "HEAD"])
+                b.check_returncode()
+                def parse_symref_head(b) -> str:
+                    out = b.stdout.decode('utf-8')
+                    """
```

Comment:
nit: we don't use docstrings inside function body like this. Consider using `#` comments and/or making it a doctest. Something like:

    def parse_symref_head(git_output: str) -> Optional[str]:
        """Extract the branch name referred by HEAD.
        Return None if the branch name cannot be detected.
        git_output is the output of 'git ls-remote --symref url HEAD'.

        >>> parse_symref_head('ref: refs/heads/main\tHEAD\nd0a9bf5e93f6f6828d4b7e3ff740c099d2ca6866\tHEAD\n')
        'main'
        >>> parse_symref_head('')
        None
        >>> parse_symref_head('ref: refs/tags/main\tHEAD\nd0a9bf5e93f6f6828d4b7e3ff740c099d2ca6866\tHEAD\n')
        None
        """

Adding `testmod("edenscm.ext.remotenames")` to `tests/test-doctest.py` would include the doctest in tests.

The function should handle cases where HEAD does not exist or is not a branch (ex. prefix is not 'refs/heads'). Some tests actually rely on non-existed HEAD so they currently break with this change.

Reply:
Hey @quark-zju, thanks for the reply! Can you share more info about doctests? I'm not familiar with how they work. 

Also what should Sapling do if parse_symref_head returns none -- is it expected to abort?

Reply:
@discentem I'll work on cleaning this up now that I have it imported on our side.

Reply:
Great! Thanks @bolinfest 😀